### PR TITLE
Explicitly link unreturned exchange orders to the original order

### PR DESCRIPTION
**Pull request by richardnuno**
This changes the way an order is determined to be an unreturned
exchange. Previously we relied on date logic. With this
change there will be an explicit link between an unreturned exchange and
it's original order.

A migration to backfill the link between the orders is provided. A
message is also surfaced in the admin to indicate that an order is an
unreturned exchange or was the original order of an unreturned exchange.

### CODE REVIEW

File: core/app/models/spree/order.rb [hunk 1]
```diff
@@ -597,15 +598,7 @@ def fully_discounted?
...
-      # created_at - 1 is a hack to ensure that this doesn't blow up on MySQL,
-      # records loaded from the DB on MySQL will have a precision of 1 second,
-      # but records in memory may still have miliseconds on them, causing this
-      # to be true where it shouldn't be.
-      #
-      # FIXME: find a better way to determine if an order is an unreturned
-      # exchange
-      shipment = self.shipments.first
-      shipment.present? ? (shipment.created_at < self.created_at - 1) : false
+      !!self.unreturned_exchange_order_id
```
Comment (philbirt) [inline core/app/models/spree/order.rb:601]: Yes. :100:

File: core/app/models/spree/order.rb [hunk 2]
```diff
@@ -120,7 +121,7 @@ def self.by_number(number)
 
     # shows completed orders first, by their completed_at date, then uncompleted orders by their created_at
     scope :reverse_chronological, -> { order('spree_orders.completed_at IS NULL', completed_at: :desc, created_at: :desc) }
-    scope :unreturned_exchange, -> { joins(:shipments).where('spree_orders.created_at > spree_shipments.created_at') }
+    scope :unreturned_exchange, -> { where.not(unreturned_exchange_order_id: nil) }
```
Comment (philbirt) [inline core/app/models/spree/order.rb:124]: Yes. :100:

File: core/db/migrate/20160203152910_backfill_unreturned_exchange_orders.rb [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    Spree::Order.complete.joins(:shipments).where('spree_orders.created_at > spree_shipments.created_at').find_each do |order|
+      return_items = Spree::ReturnItem.where(exchange_inventory_unit_id: order.inventory_unit_ids)
+      if return_items.any?
+        ra_ids = return_items.pluck(:return_authorization_id).compact
+        cr_ids = return_items.pluck(:customer_return_id).compact
+        ra_order_ids = Spree::ReturnAuthorization.where(id: ra_ids).pluck(:order_id)
+        cr_order_ids = Spree::CustomerReturn.where(id: cr_ids).map(&:order_id)
+
+        # Remove the unreturned exchange order id in case it has exchanges itself
+        order_ids = (ra_order_ids + cr_order_ids).compact.uniq - [order.id]
```
Comment (philbirt) [inline core/db/migrate/20160203152910_backfill_unreturned_exchange_orders.rb:14]: This makes me scared, but that's probably a normal reaction

File: backend/app/views/spree/admin/orders/_carton.html.erb [hunk 4]
```diff
@@ -69,17 +69,6 @@
         </td>
       </tr>
 
-      <% if carton.orders.size > 1 %>
```
Comment (philbirt) [inline backend/app/views/spree/admin/orders/_carton.html.erb:72]: Nice

**Comment by philbirt** [comment]
Overall looks great. That migration is a little scary, but given my experience with this stuff, its probably what has to be done. We should spot-check heavily. How many orders get updated from this?

File: core/db/migrate/20160203152910_backfill_unreturned_exchange_orders.rb (hunk 3 above)
Comment (richardnuno) [inline core/db/migrate/20160203152910_backfill_unreturned_exchange_orders.rb:14]: Yeah, that's why I added the check to ensure the result is always only one order.

**Comment by richardnuno** [comment]
~3000 orders will get the explicit link

**Comment by richardnuno** [comment]
Closing this to open against master

**Pull request closed without merging**